### PR TITLE
[tune] better signature check for `tune.sample_from`

### LINKED ISSUE #13009: [tune] sample_from validation fails with functools.partial and callables with kwargs
CC @richardliaw 

(labeling as bug because functools.partial and callables with kwargs should definitely be supported [IMO] and there's no indication in the docs that they don't work)

### What is the problem?
https://github.com/ray-project/ray/blob/4caa6c6d7806d1e406322bfb50fa01072f63985e/python/ray/tune/sample.py#L317-L320

In `tune.sample_from` we have a check to see if the passed callable has either 0 or 1 arguments. This fails for `functools.partial` and callables that have argument with default values. 

### Reproduction (REQUIRED)
```python
# untested; you get the idea
from ray import tune


def trainable(config):
    return config["a"]

def get_a(spec, base=2):
    return base/(spec.b**2)

config = {
    "b": tune.uniform(0, 10),
    "a": tune.sample_from(get_a),
}
tune.run(trainable, config=config)
```

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

#### Make `spec` a positional argument
i.e., instruct users to make this their `sample_from` signature:
```python
def func(spec, /, arg, kwarg=None):
    pass
```
then we could change the check to `if len(frozenset(name for name, param in signature(func).parameters.items() if param.kind == Parameter.POSITIONAL_ONLY)) > 1`

pros:
- we get to keep the check

cons:
- seems a little unintuitive for users to be required to do this

#### Remove the check
pros:
- this GH issue is resolved

cons:
- uglier error message is user passes function with an incorrect signature?
    - we can `except NameError` to print a prettier error message

### PR DESCRIPTION
**Pull request by krfricke**
## Related issue number

Closes #13009

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/sample.py [hunk 1]
```diff
@@ -347,11 +346,23 @@ def sample(self,
...
 
     def __init__(self, func: Callable):
-        if len(signature(func).parameters) > 1:
-            raise ValueError(
-                "The function passed to a `Function` parameter must accept "
-                "either 0 or 1 parameters.")
-
+        sig = signature(func)
+
+        pass_spec = True
```
Comment (sumanthratna) [inline python/ray/tune/sample.py:351]:
```suggestion
        pass_spec = True  # whether we should pass `spec` when calling `func`
```

Comment (sumanthratna) [reply python/ray/tune/sample.py:351]: (feel free to ignore btw)

**Review by sumanthratna — APPROVED** [review]
thanks @krfricke it looks great! Added one tiny suggestion because at first I thought `pass_spec` was for whether the func is a valid domain instead of whether `spec` should be passed to `func` (although I will admit the variable name is pretty clear here)

**Comment by krfricke** [comment]
I guess more documentation is always better, so committed your suggestion. Thanks!

**Comment by krfricke** [comment]
This should be ready to merge @richardliaw

**Pull request merged**